File: lua/mpp/server/html.lua
local assets = require('mpp.server.html_assets.gen')
local base64url = require('mpp.util.base64url')
local json = require('mpp.util.json')

local M = {}

local function html_escape(str)
  if type(str) ~= 'string' then
    return ''
  end
  return str
    :gsub('&', '&amp;')
    :gsub('<', '&lt;')
    :gsub('>', '&gt;')
    :gsub('"', '&quot;')
    :gsub("'", '&#39;')
end

function M.accepts_html(accept_header)
  if type(accept_header) ~= 'string' then
    return false
  end
  return accept_header:find('text/html', 1, true) ~= nil
end

function M.is_service_worker_request(uri_args)
  if type(uri_args) ~= 'table' then
    return false
  end
  return uri_args.__mpp_worker ~= nil
end

function M.service_worker_js()
  return assets.service_worker_js
end

function M.challenge_to_html(challenge, rpc_url)
  local plain = {
    id = challenge.id,
    realm = challenge.realm,
    method = challenge.method,
    intent = challenge.intent,
    request = challenge.request:raw(),
    expires = challenge.expires,
    description = challenge.description,
    digest = challenge.digest,
    opaque = challenge.opaque and challenge.opaque:raw() or nil,
  }

  local challenge_json = json.encode(plain)

  -- Decode the base64url request field to extract network from methodDetails.
  local network = 'mainnet-beta'
  local decoded_payload, decode_err = base64url.decode(plain.request)
  if decoded_payload then
    local ok, request_data = pcall(json.decode, decoded_payload)
    if ok and type(request_data) == 'table' then
      local method_details = request_data.methodDetails
      if type(method_details) == 'table' and type(method_details.network) == 'string' then
        network = method_details.network
      end
    end
  end

  local test_mode = (network == 'devnet' or network == 'localnet')

  local embedded_data = {
    challenge = plain,
    network = network,
    rpcUrl = rpc_url,
    testMode = test_mode,
  }
  local embedded_json = json.encode(embedded_data)

  local parts = {
    '<!DOCTYPE html>',
    '<html lang="en">',
    '<head>',
    '<meta charset="utf-8">',
    '<meta name="viewport" content="width=device-width, initial-scale=1">',
    '<title>Payment Required</title>',
    '<style>',
    'body { font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, sans-serif; margin: 0; padding: 20px; background: #f7fafc; color: #1a202c; }',
    'pre { background: #edf2f7; padding: 16px; border-radius: 8px; overflow-x: auto; font-size: 13px; max-width: 600px; margin: 20px auto; }',
    '</style>',
    '</head>',
    '<body>',
    '<details style="max-width:600px;margin:0 auto 20px">',
    '<summary style="cursor:pointer;color:#718096;font-size:14px">Challenge details</summary>',
    '<pre>' .. html_escape(challenge_json) .. '</pre>',
    '</details>',
    '<div id="root"></div>',
    -- JSON inside <script type="application/json"> is not parsed as HTML.
    -- json.encode already escapes special chars in string values.
    '<script type="application/json" id="__MPP_DATA__">' .. embedded_json .. '</script>',
    '<script>' .. assets.payment_ui_js .. '</script>',
    '</body>',
    '</html>',
  }

  return table.concat(parts, '\n')
end

return M

File: typescript/packages/mpp/src/server/html.ts
/**
 * HTML payment link support for browser-based payments.
 *
 * When enabled, 402 responses can be rendered as interactive HTML payment pages.
 * The page embeds a bundled Solana payment UI that handles wallet connection,
 * transaction signing, and credential submission via a service worker.
 *
 * @example
 * ```ts
 * import { html } from '@solana/mpp/server/html'
 *
 * // In your HTTP handler:
 * if (html.acceptsHtml(request.headers.get('accept'))) {
 *   if (html.isServiceWorkerRequest(request.url)) {
 *     return new Response(html.serviceWorkerJs(), {
 *       headers: {
 *         'Content-Type': 'application/javascript',
 *         'Service-Worker-Allowed': '/',
 *       },
 *     })
 *   }
 *   const challenge = await mppx.charge({ amount: '10000', currency: 'USDC' })(request)
 *   if (challenge.status === 402) {
 *     return html.respondWithPaymentPage(challenge.challenge)
 *   }
 * }
 * ```
 */

import { PAYMENT_UI_JS, SERVICE_WORKER_JS } from './html-assets.gen.js';

/** Query parameter that triggers serving the service worker JS. */
export const SERVICE_WORKER_PARAM = '__mpp_worker';

/** ID of the embedded data script element. */
const DATA_ELEMENT_ID = '__MPP_DATA__';

/** Check if an HTTP Accept header value includes text/html. */
export function acceptsHtml(accept: string | null | undefined): boolean {
    if (!accept) return false;
    return accept.split(',').some((part) => part.trim().startsWith('text/html'));
}

/** Check if a URL contains the service worker query parameter. */
export function isServiceWorkerRequest(url: string): boolean {
    return url.includes(SERVICE_WORKER_PARAM);
}

/** Returns the service worker JavaScript content. */
export function serviceWorkerJs(): string {
    return SERVICE_WORKER_JS as string;
}

/** Returns the payment UI JavaScript content (for advanced use). */
export function paymentUiJs(): string {
    return PAYMENT_UI_JS as string;
}

/** HTML-escape a string to prevent XSS. */
function escapeHtml(s: string): string {
    return s
        .replace(/&/g, '&amp;')
        .replace(/</g, '&lt;')
        .replace(/>/g, '&gt;')
        .replace(/"/g, '&quot;')
        .replace(/'/g, '&#x27;');
}

export interface ChallengeData {
    description?: string;
    digest?: string;
    expires?: string;
    id: string;
    intent: string;
    method: string;
    opaque?: string;
    realm: string;
    request: string;
}

export interface PaymentPageOptions {
    /** The payment challenge to embed in the page. */
    challenge: ChallengeData;
    /** Solana network (mainnet-beta, devnet, localnet). */
    network: string;
    /** RPC URL for the payment UI to use. */
    rpcUrl: string;
}

/**
 * Render a payment challenge as a self-contained HTML payment page.
 *
 * The returned HTML includes the challenge data and inlined payment UI script.
 * All user-controlled values are HTML-escaped to prevent XSS.
 */
export function challengeToHtml(options: PaymentPageOptions): string {
    const { challenge, network, rpcUrl } = options;
    const testMode = network === 'devnet' || network === 'localnet';
    const challengeJson = JSON.stringify(challenge, null, 2);

    const embeddedData = JSON.stringify({
        challenge,
        network,
        rpcUrl,
        testMode,
    });

    return `<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Payment Required</title>
<style>
body { font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, sans-serif; margin: 0; padding: 20px; background: #f7fafc; color: #1a202c; }
pre { background: #edf2f7; padding: 16px; border-radius: 8px; overflow-x: auto; font-size: 13px; max-width: 600px; margin: 20px auto; }
</style>
</head>
<body>
<details style="max-width:600px;margin:0 auto 20px">
<summary style="cursor:pointer;color:#718096;font-size:14px">Challenge details</summary>
<pre>${escapeHtml(challengeJson)}</pre>
</details>
<div id="root"></div>
<script type="application/json" id="${DATA_ELEMENT_ID}">${embeddedData}</script>
<script>${PAYMENT_UI_JS}</script>
</body>
</html>`;
}

/**
 * Create a full 402 HTML Response for a payment challenge.
 *
 * Sets Content-Type, Content-Security-Policy, WWW-Authenticate, and Cache-Control headers.
 */
export function respondWithPaymentPage(
    options: PaymentPageOptions & { wwwAuthenticate: string },
): Response {
    const html = challengeToHtml(options);
    return new Response(html, {
        headers: {
            'Cache-Control': 'no-store',
            'Content-Security-Policy':
                "default-src 'self'; script-src 'unsafe-inline'; style-src 'unsafe-inline'; connect-src *; worker-src 'self'",
            'Content-Type': 'text/html; charset=utf-8',
            'WWW-Authenticate': options.wwwAuthenticate,
        },
        status: 402,
    });
}
